### ISSUE TITLE
Translation export fails when "Select all templates" is used

### ISSUE THREAD

**Issue by trendspotter** [sahana/eden#1468]
Steps to reproduce:
1. Go to `/eden/admin/translate/create`
2. Check the *Select all templates (All modules included)* checkbox
3. Enter a new language code (eg. *xx*)
4. Click *Submit*

Expected result:
XLS file with translation strings will be offered for download.

Actual result:
Error 500 + following traceback
```
Traceback (most recent call last):
  File "/srv/web2py/gluon/restricted.py", line 219, in restricted
    exec(ccode, environment)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1164, in <module>
  File "/srv/web2py/gluon/globals.py", line 419, in <lambda>
    self._caller = lambda f: f()
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1087, in translate
    output = s3_rest_controller("translate", "language")
  File "/srv/web2py/applications/eden/models/00_utils.py", line 243, in s3_rest_controller
    output = r(**attr)
  File "applications/eden/modules/s3/s3rest.py", line 693, in __call__
    output = postprocess(self, output)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 851, in postp
    output = X.export_file(code, modlist, [], filetype, all_template_flag)
  File "applications/eden/modules/s3/s3translate.py", line 1112, in export_file
    return self.write_xls(Strings, langcode)
  File "applications/eden/modules/s3/s3translate.py", line 1290, in write_xls
    wbk.save(output)
  File "/usr/lib/python2.7/site-packages/xlwt/Workbook.py", line 710, in save
    doc.save(filename_or_stream, self.get_biff_data())
  File "/usr/lib/python2.7/site-packages/xlwt/Workbook.py", line 674, in get_biff_data
    shared_str_table   = self.__sst_rec()
  File "/usr/lib/python2.7/site-packages/xlwt/Workbook.py", line 636, in __sst_rec
    return self.__sst.get_biff_record()
  File "/usr/lib/python2.7/site-packages/xlwt/BIFFRecords.py", line 77, in get_biff_record
    self._add_to_sst(s)
  File "/usr/lib/python2.7/site-packages/xlwt/BIFFRecords.py", line 92, in _add_to_sst
    u_str = upack2(s, self.encoding)
  File "/usr/lib/python2.7/site-packages/xlwt/UnicodeUtils.py", line 55, in upack2
    raise Exception('String longer than 32767 characters')
Exception: String longer than 32767 characters
```

Examining the XLS during creation shows that there is some weird recursion at play which causes the export to repeat the string location and eventually hits the character limit given by *xlwt*.

**Comment by hallamoore** [comment]
@trendspotter are you still seeing this bug? I followed your steps to reproduce, but I was able to download the XLS file. It did take a long time to process after clicking submit, but it eventually succeeded. (The url appears to have changed slightly, I went to `eden/admin/translate/create?opt=1`)

**Comment by AlexanderLaughlin** [comment]
@trendspotter @hallamoore is this still an open issue?

**Comment by flavour** [comment]
I have been unable to reproduce this...possibly it's due to an older version of the xlwt library?

**Comment by trendspotter** [comment]
Currently, I'm also unable to reproduce as the export, following the exact same steps as in the original report, fails with another exception.

```
Traceback (most recent call last):
  File "/srv/web2py/gluon/restricted.py", line 219, in restricted
    exec(ccode, environment)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1271, in <module>
  File "/srv/web2py/gluon/globals.py", line 421, in <lambda>
    self._caller = lambda f: f()
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1194, in translate
    output = s3_rest_controller("translate", "language")
  File "/srv/web2py/applications/eden/models/00_utils.py", line 244, in s3_rest_controller
    output = r(**attr)
  File "/srv/web2py/applications/eden/modules/s3/s3rest.py", line 689, in __call__
    output = postprocess(self, output)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 961, in postp
    output = X.export_file(code, modlist, [], filetype, all_template_flag)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1060, in export_file
    NewStrings += get_strings_by_module(mod)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 125, in get_strings_by_module
    tmpstr = R.read_html_js(f)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 749, in read_html_js
    for line in html_js_file:
  File "/usr/lib/python3.8/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 526: ordinal not in range(128)
```

The offending file is `eden/static/scripts/gis/openlayers/lib/OpenLayers/Lang/br.js` which indeed contains non-ascii characters.

**Comment by flavour** [comment]
I've not had time to test it yet, but I suspect this commit should fix that:
https://github.com/sahana/eden/commit/11709a2720e53744f6aa402d8b0a5ed5b948f614

**Comment by trendspotter** [comment]
> I've not had time to test it yet, but I suspect this commit should fix that:
> [11709a2](https://github.com/sahana/eden/commit/11709a2720e53744f6aa402d8b0a5ed5b948f614)

That *did* help, however we're not there just yet.

```
Traceback (most recent call last):
  File "/srv/web2py/gluon/restricted.py", line 219, in restricted
    exec(ccode, environment)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1271, in <module>
  File "/srv/web2py/gluon/globals.py", line 421, in <lambda>
    self._caller = lambda f: f()
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1194, in translate
    output = s3_rest_controller("translate", "language")
  File "/srv/web2py/applications/eden/models/00_utils.py", line 244, in s3_rest_controller
    output = r(**attr)
  File "/srv/web2py/applications/eden/modules/s3/s3rest.py", line 689, in __call__
    output = postprocess(self, output)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 961, in postp
    output = X.export_file(code, modlist, [], filetype, all_template_flag)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1060, in export_file
    NewStrings += get_strings_by_module(mod)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 122, in get_strings_by_module
    tmpstr = findstr(f, "ALL", modlist)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 653, in findstr
    fileContent = "%s\n" % fileContent.replace("\r", "")
TypeError: a bytes-like object is required, not 'str'
```

**Comment by flavour** [comment]
Again, am not able to test right now, but this may help...reverted part of the earlier fix & also fixed an obvious need for BytesIO instead of StringIO
https://github.com/sahana/eden/commit/a1f072de25378e3fdcbf1976b324df9f0fcd3021

**Comment by trendspotter** [comment]
Whac-a-mole :)

```
Traceback (most recent call last):
  File "/srv/web2py/gluon/restricted.py", line 219, in restricted
    exec(ccode, environment)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1271, in <module>
  File "/srv/web2py/gluon/globals.py", line 421, in <lambda>
    self._caller = lambda f: f()
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1194, in translate
    output = s3_rest_controller("translate", "language")
  File "/srv/web2py/applications/eden/models/00_utils.py", line 244, in s3_rest_controller
    output = r(**attr)
  File "/srv/web2py/applications/eden/modules/s3/s3rest.py", line 689, in __call__
    output = postprocess(self, output)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 961, in postp
    output = X.export_file(code, modlist, [], filetype, all_template_flag)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1060, in export_file
    NewStrings += get_strings_by_module(mod)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 125, in get_strings_by_module
    tmpstr = R.read_html_js(f)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 751, in read_html_js
    occur = findall(line)
TypeError: cannot use a string pattern on a bytes-like object
```

**Comment by trendspotter** [comment]
Also note that some strings are shown as byte-strings, e.g. the info/alert message pop-ups. When I log out, the message shows `b'Logged out'`. This gets worse then the strings are localized and use non-ASCII characters as they are shown `\x` encoded. This is unrelated to the original topic of this issue, but it's evident that we're now hunting down encoding problems, so I'm just mentioning it for completeness.

**Comment by flavour** [comment]
I saw that with an older web2py...are you running the same hash as our
install scripts?

**Comment by trendspotter** [comment]
Aaah, yes. You're right. We were still using 2.18.3 (6128d03), while the scripts list 2.18.5 (59700b8). After update, the problem with alerts went away.

The original problem with translation exports still stands. I've checked the last 3 eden commits against 2.18.5, just in case there is any difference in relation to this ticket, but nope, in all cases, the errors and tracebacks remain the same.

**Comment by flavour** [comment]
Thanks for the detailed checking :)
I took some time to get it working in Py 2.7, 3.7 & 3.8:
https://github.com/sahana/eden/commit/fd6a5f48b2a2cf04b8649a8f6e387acc6afee003

Will close Issue, but feel free to reopen if there are still problems at your end

**Comment by trendspotter** [comment]
I'm a demanding customer :) Seems like we're really close as the export fails after a few seconds and not immediately, but it fails nevertheless.

```
Traceback (most recent call last):
  File "/srv/web2py/gluon/restricted.py", line 219, in restricted
    exec(ccode, environment)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1271, in <module>
  File "/srv/web2py/gluon/globals.py", line 421, in <lambda>
    self._caller = lambda f: f()
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1194, in translate
    output = s3_rest_controller("translate", "language")
  File "/srv/web2py/applications/eden/models/00_utils.py", line 244, in s3_rest_controller
    output = r(**attr)
  File "/srv/web2py/applications/eden/modules/s3/s3rest.py", line 689, in __call__
    output = postprocess(self, output)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 961, in postp
    output = X.export_file(code, modlist, [], filetype, all_template_flag)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1083, in export_file
    NewStrings += R.get_database_strings(all_template_flag)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 905, in get_database_strings
    data = read_csv(csv_path)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1130, in read_csv
    for row in transReader:
  File "/usr/lib/python3.8/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 3849: ordinal not in range(128)
```

The offending file this time seems to be `eden/modules/templates/RMSAmericas/hrm_course.csv`. (Note that this is the *all templates* export.)

Without *all templates*, I'm getting another one

```
Traceback (most recent call last):
  File "/srv/web2py/gluon/restricted.py", line 219, in restricted
    exec(ccode, environment)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1271, in <module>
  File "/srv/web2py/gluon/globals.py", line 421, in <lambda>
    self._caller = lambda f: f()
  File "/srv/web2py/applications/eden/controllers/admin.py", line 1194, in translate
    output = s3_rest_controller("translate", "language")
  File "/srv/web2py/applications/eden/models/00_utils.py", line 244, in s3_rest_controller
    output = r(**attr)
  File "/srv/web2py/applications/eden/modules/s3/s3rest.py", line 689, in __call__
    output = postprocess(self, output)
  File "/srv/web2py/applications/eden/controllers/admin.py", line 961, in postp
    output = X.export_file(code, modlist, [], filetype, all_template_flag)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1114, in export_file
    return self.write_xls(Strings, langcode)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1279, in write_xls
    d2 = string_escape(d2)
  File "/srv/web2py/applications/eden/modules/s3/s3translate.py", line 1272, in string_escape
    return(s.encode("latin1")         # To bytes, required by 'unicode-escape'
UnicodeEncodeError: 'latin-1' codec can't encode character '\u010d' in position 2: ordinal not in range(256)
```

but I haven't pinpointed yet where the input comes from in this case, so it *may* be caused by our template.

Apart from that, there are a few warnings. Likely nothing interesting.

```
2020-02-09 13:58:20 S3LOG WARNING: Can't find this deployment_setting, maybe a crud.settings: submit_button
2020-02-09 13:58:23 S3LOG WARNING: /srv/web2py/applications/eden/static/scripts/gis/proj4js/lib/proj4js.js is in LATIN-1 encoding not UTF-8
2020-02-09 13:58:23 S3LOG WARNING: /srv/web2py/applications/eden/static/scripts/gis/proj4js/lib/proj4js-combined.js is in LATIN-1 encoding not UTF-8
2020-02-09 13:58:24 S3LOG WARNING: /srv/web2py/applications/eden/static/scripts/ext/src/locale/ext-lang-pt_PT.js is in LATIN-1 encoding not UTF-8
2020-02-09 13:58:24 S3LOG WARNING: /srv/web2py/applications/eden/static/scripts/ext/src/locale/ext-lang-he.js is in LATIN-1 encoding not UTF-8
2020-02-09 13:58:26 S3LOG WARNING: /srv/web2py/applications/eden/static/scripts/trunk8.min.js is in LATIN-1 encoding not UTF-8
2020-02-09 13:58:28 S3LOG WARNING: Module 'error' doesn't exist!
```

(GitHub rules won't allow me to reopen the issue as I'm not a collaborator and I haven't closed the issue myself.)

**Comment by flavour** [comment]
I get the WARNINGs too.
Which language are you using for the export?
Python 3.8.1?
XLS or PO?

I did test with all modules, so will be reading that file

**Comment by trendspotter** [comment]
Language "ar" (the first one without even selecting any other)
Python 3.8.1
XLS

Basically I just navigate to /eden/admin/translate, check the *Select all templates* checkbox and click on *Submit*. However I'm getting both exceptions on any language and export file type.

I've just spun another container with the undesecrated default template and I'm getting still the same.

**Comment by trendspotter** [comment]
Hmm. That` string_escape` function in s3translate seems like something that has been probably needed historically but now it can be optimized out. Following patch fixes both problems for me (not submitting a PR because I'm not sure and don't want to break both the code *and* the discussion)

```diff
diff --git a/modules/s3/s3translate.py b/modules/s3/s3translate.py
index c45be060d..c773bf415 100644
--- a/modules/s3/s3translate.py
+++ b/modules/s3/s3translate.py
@@ -1125,7 +1125,7 @@ class Strings(object):

         data = []
         dappend = data.append
-        f = open(fileName, "r") # , newline=""
+        f = open(fileName, "r", encoding="utf-8") # , newline=""
         transReader = csv.reader(f)
         for row in transReader:
             dappend(row)
@@ -1263,15 +1263,11 @@ class Strings(object):
         row_num = 1

         if PY2:
-            def string_escape(s, encoding="utf-8"):
-                return(s.decode("string-escape")
-                        .decode("utf-8"))
+            def string_escape(s):
+                return s.decode("utf-8")
         else:
-            def string_escape(s, encoding="utf-8"):
-                return(s.encode("latin1")         # To bytes, required by 'unicode-escape'
-                        .decode("unicode-escape") # Perform the actual octal-escaping decode
-                        .encode("latin1")         # 1:1 mapping back to bytes
-                        .decode(encoding))        # Decode original encoding
+            def string_escape(s):
+                return s

         # Write the data to spreadsheet
         for (loc, d1, d2) in Strings:
```

That `.encode("latin1")` is wrong on all accounts as any string with characters not fitting to latin-1 space will inherently fail there. The whole point of the function seems to be to convert the string to utf-8  (or any parameter give as `encoding`, which is not used by any caller anywhere in the code) which is safe to be printed out to the file. In python3 everything it utf-8 by default, so there seems to be little point in trying to convert it again (and if so, certainly not via latin1).

Also in python2 decoding into string-escape and then *again* to utf-8 seem to be superfluous. The encoding parameter is even ignored completely there.

**Comment by flavour** [comment]
If we don't need to string-escape/unicode-escape, then we can use s3_str as that's what your 'string_escape' does...however there must have been a reason why the original programmer put in string-escape, which in Py3 has become unicode-escape

**Comment by flavour** [comment]
This is where I got that code from:
https://stackoverflow.com/questions/14820429/how-do-i-decodestring-escape-in-python3

**Comment by trendspotter** [comment]
Well in that case that 'latin1' literal could be omitted from both .decode()s, in which case python will use utf-8 as default and the non-latin1 characters will work too.
```
>>> '\\u011b\\u0161\\u010d\\u0159\\u017e'.encode('latin1').decode('unicode-escape').encode('latin1').decode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'latin-1' codec can't encode characters in position 0-4: ordinal not in range(256)
```
vs
```
>>> '\\u011b\\u0161\\u010d\\u0159\\u017e'.encode().decode('unicode-escape').encode().decode('utf-8')
'ěščřž'
```

However this expects that the input string will be double-escaped (double-backslashes passed as part of the string), which doesn't seem to be the case anywhere. In case of single-escaping, the explicit conversion is not only unnecessary (python will do it implicitly) but outright harmful.
```
>>> 'ěščřž'.encode('latin1').decode('unicode-escape').encode('latin1').decode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'latin-1' codec can't encode characters in position 0-4: ordinal not in range(256)
```
```
>>> 'ěščřž'.encode().decode('unicode-escape').encode().decode('utf-8')
'Ä\x9bÅ¡Ä\x8dÅ\x99Å¾'
```

As I have mentioned, the patch above got me successful export with meaningful data. Any other combination mentioned above either didn't work or ended up with garbled and unusable strings.

**Comment by trendspotter** [comment]
The longer I look on the write_xls method (and it's been almost an 2 hours :) ), the more I think that the string_escape is some relic of a bygone era and it's entirely unnecessary for python3. The XLS is created as an UTF-8 file, so all strings passed to it are expected to be utf-8. `Strings` are passed from `export_file()` method and are populated as strings (i.e. not double-escaped sequences or whatever). So there is no point on transforming them back and forth in order to get them written as utf-8 strings to XLS. They already are utf-8 strings to begin with. Moreover, no such explicit conversion is done for PO/CSV files.

In fact, the whole opening files as binary and explicitly decoding them seems like a bad idea. Let me try to step back a bit and try a fix now when I've started into the translation code long enough and have some idea how it's supposed to work. :)

**Comment by flavour** [comment]
I have discovered what string-escape did in the Python 2 world: it converts accented characters to ascii equivalents.
I think I have a solution for Py3...but I need to go to Panto rehearsal before my test run completes:
unicodedata.normalize("NFD", s).encode("ascii", "ignore").decode("utf-8")

**Comment by trendspotter** [comment]
1) I don't see how would that be useful. Translator would want to see unaltered original translations, not ascii-safe.

2) It doesn't seem to do that on python 2.7

```
>>> 'ěščřž'.decode("string-escape").decode("utf-8")
u'\u011b\u0161\u010d\u0159\u017e'

>>> 'ěščřž'.decode("string-escape")
'\xc4\x9b\xc5\xa1\xc4\x8d\xc5\x99\xc5\xbe'

>>> 'ěščřž'.decode("string-escape").decode("utf-8")
u'\u011b\u0161\u010d\u0159\u017e'

>>> 'ěščřž'.decode("utf-8")
u'\u011b\u0161\u010d\u0159\u017e'
```

3) It seems to be some internal undocumented feature - https://bugs.python.org/issue30588 :)

4) unicodedata.normalize requires additional import and also requires the parameter to be already unicode string on python2

```
>>> unicodedata.normalize("NFD", 'ěščřž').encode("ascii", "ignore").decode("utf-8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: normalize() argument 2 must be unicode, not str

>>> unicodedata.normalize("NFD", u'ěščřž').encode("ascii", "ignore").decode("utf-8")
u'escrz'
```

I don't think that's the right solution either. Anyway, let me have a bit of fun with this... :)

**Comment by flavour** [comment]
> 1. I don't see how would that be useful. Translator would want to see unaltered original translations, not ascii-safe.

Yes, have no idea either why this would be wanted but it does seem that's what it was there for as:

> 2. It doesn't seem to do that on python 2.7
> >>> 'ěščřž'.decode("string-escape").decode("utf-8")
> u'\u011b\u0161\u010d\u0159\u017e'

It does for me:
Python 2.7.16 (v2.7.16:413a49145e, Mar  4 2019, 01:30:55) [MSC v.1500 32 bit (Intel)] on win32
>>> 'ěščřž'.decode('string-escape').decode("utf-8")
u'escrz'


> 3. It seems to be some internal undocumented feature - https://bugs.python.org/issue30588 :)

I am not using escape_decode?

> 4. unicodedata.normalize requires additional import and also requires the parameter to be already unicode string on python2

I was planning to leave the Py2 version as-is...just use that version for Py3

> I don't think that's the right solution either. Anyway, let me have a bit of fun with this... :)

OK, thanks, will leave for tonight

**Comment by flavour** [comment]
Well, I wanted to leave something unbroken...just not doing string_escape at all seems fine for now (which is pretty much what you said above although what you had broke 2.7).
Will leave open whilst you play

**Comment by trendspotter** [comment]
It mostly works now, so I have just submitted #1533 for the finishing works.

The better part of what I did yesterday revolved around *not* using `string_escape` and trying to handle the files with correct encoding in py3. One iteration I had was that I've even added a dictionary to s3compat, something in lines:
```
if PY2:
    PY3ENCODING={}
else:
    PY3ENCODING={'encoding':'utf-8'}
```
and then passed as keyword arguments to the respective opens as 
```
open(filename, mode, **PY3ENCODING)
```
I'm mentioning this as I fear that something like this may eventually be needed as I expect the encoding problems to be widespread in all other modules importing/exporting any files.

> It does for me:
> Python 2.7.16 (v2.7.16:413a49145e, Mar 4 2019, 01:30:55) [MSC v.1500 32 bit (Intel)] on win32
> ```
> 'ěščřž'.decode('string-escape').decode("utf-8")
> u'escrz'
> ```
Yeah. This is probably why it's undocumented. This seems to be somehow locale-specific or even OS-specific. To "asciify" a string, I have been using the unicodedata at least since 2013 (even wrote a blogpost about that :) )

> I am not using escape_decode?

That issue explicitly mentions `string-escape` in 2 comments. I guess `escape_decode` is just a common name for the feature and the function using it. One comment also mention the exact same StackOverflow question you've linked before.
